rln/src/iden3calc/storage.rs

```diff
@@ -0,0 +1,493 @@
+            proto::DuoOp::Sub => Operation::Sub,
+            proto::DuoOp::Pow => Operation::Pow,
+            proto::DuoOp::Idiv => Operation::Idiv,
+            proto::DuoOp::Mod => Operation::Mod,
+            proto::DuoOp::Eq => Operation::Eq,
+            proto::DuoOp::Neq => Operation::Neq,
+            proto::DuoOp::Lt => Operation::Lt,
+            proto::DuoOp::Gt => Operation::Gt,
+            proto::DuoOp::Leq => Operation::Leq,
+            proto::DuoOp::Geq => Operation::Geq,
+            proto::DuoOp::Land => Operation::Land,
+            proto::DuoOp::Lor => Operation::Lor,
+            proto::DuoOp::Shl => Operation::Shl,
+            proto::DuoOp::Shr => Operation::Shr,
+            proto::DuoOp::Bor => Operation::Bor,
+            proto::DuoOp::Band => Operation::Band,
+            proto::DuoOp::Bxor => Operation::Bxor,
+        }
+    }
+}
+
+impl From<proto::TresOp> for graph::TresOperation {
+    fn from(value: proto::TresOp) -> Self {
+        match value {
+            proto::TresOp::TernCond => TresOperation::TernCond,
+        }
+    }
+}
+
+pub fn serialize_witnesscalc_graph<T: Write>(
+    mut w: T,
+    nodes: &Vec<graph::Node>,
+    witness_signals: &Vec<usize>,
+    input_signals: &InputSignalsInfo,
+) -> std::io::Result<()> {
+    let mut ptr = 0usize;
+    w.write_all(WITNESSCALC_GRAPH_MAGIC).unwrap();
+    ptr += WITNESSCALC_GRAPH_MAGIC.len();
+
+    w.write_u64::<LittleEndian>(nodes.len() as u64)?;
+    ptr += 8;
+
+    let metadata = proto::GraphMetadata {
+        witness_signals: witness_signals
+            .iter()
+            .map(|x| *x as u32)
+            .collect::<Vec<u32>>(),
+        inputs: input_signals
+            .iter()
+            .map(|(k, v)| {
+                let sig = proto::SignalDescription {
+                    offset: v.0 as u32,
+                    len: v.1 as u32,
+                };
+                (k.clone(), sig)
+            })
+            .collect(),
+    };
+
+    // capacity of buf should be enough to hold the largest message + 10 bytes
+    // of varint length
+    let mut buf = Vec::with_capacity(metadata.encoded_len() + MAX_VARINT_LENGTH);
+
+    for node in nodes {
+        let node_pb = proto::Node {
+            node: Some(proto::node::Node::from(node)),
+        };
+
+        assert_eq!(buf.len(), 0);
+        node_pb.encode_length_delimited(&mut buf)?;
+        ptr += buf.len();
+
+        w.write_all(&buf)?;
+        buf.clear();
+    }
+
+    metadata.encode_length_delimited(&mut buf)?;
+    w.write_all(&buf)?;
+    buf.clear();
+
+    w.write_u64::<LittleEndian>(ptr as u64)?;
+
+    Ok(())
+}
+
+fn read_message_length<R: Read>(rw: &mut WriteBackReader<R>) -> std::io::Result<usize> {
+    let mut buf = [0u8; MAX_VARINT_LENGTH];
+    let bytes_read = rw.read(&mut buf)?;
+    if bytes_read == 0 {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::UnexpectedEof,
+            "Unexpected EOF",
```

Comment:
Probably better to have these literals by name, reuse them and put them in a different file. Even better would be a more holistic approach to error handling, but that could be topic of another PR

Reply:
No need in this, because the iden3 code contains only 2 such literals, which are used only 3 times, and the original zerokit code uses the same approach to error handling.
![изображение](https://github.com/user-attachments/assets/25d37a91-8dc2-405a-8693-43261ac82142)

Reply:
I still think that processing is necessary because, if you pay attention, all external functions that are called by rln have this handling and this one don't leave utils lib in this form 

Moreover, as far as I can see you use unwrap in all calls, which is not very good either: https://github.com/vacp2p/zerokit/blob/d8232a5b6eb89aba836c83531aa8bba55fe7f770/rln/src/iden3calc.rs#L24